akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala

```diff
@@ -0,0 +1,685 @@
+    val stageModule = mod.asInstanceOf[GraphStageModule[Shape, Any]]
+    val matAndLogic = stageModule.stage.createLogicAndMaterializedValue(attributes)
+    val logic = matAndLogic._1
+    logics.add(logic)
+    logic.stageId = logics.size() - 1
+    matAndLogic
+  }
+
+  def conn(slot: Int): Connection = {
+    maxConnections = math.max(slot, maxConnections)
+    val c = connections(slot)
+    if (c ne null) c
+    else {
+      val c2 = new Connection(0, 0, null, 0, null, null, null)
+      connections(slot) = c2
+      c2
+    }
+  }
+
+  def outConn(): Connection = {
+    val connection = new Connection(0, 0, null, 0, null, null, null)
+    outConnections ::= connection
+    connection
+  }
+
+  override def assignPort(in: InPort, slot: Int, logic: GraphStageLogic): Unit = {
+    val connection = conn(slot)
+    connection.inOwner = logic
+    connection.id = slot
+    connection.inOwnerId = logic.stageId
+    connection.inHandler = logic.handlers(in.id).asInstanceOf[InHandler]
+    logic.portToConn(in.id) = connection
+  }
+
+  override def assignPort(out: OutPort, slot: Int, logic: GraphStageLogic): Unit = {
+    val connection = conn(slot)
+    connection.outOwner = logic
+    connection.id = slot
+    connection.outOwnerId = logic.stageId
+    connection.outHandler = logic.handlers(logic.inCount + out.id).asInstanceOf[OutHandler]
+    logic.portToConn(logic.inCount + out.id) = connection
+  }
+
+  override def createPublisher(out: OutPort, logic: GraphStageLogic): Publisher[Any] = {
+    val boundary = new ActorOutputBoundary(shell, out.toString)
+    logics.add(boundary)
+    boundary.stageId = logics.size() - 1
+
+    val connection = outConn()
+    boundary.portToConn(boundary.in.id) = connection
+    connection.inHandler = boundary.handlers(0).asInstanceOf[InHandler]
+    connection.inOwner = boundary
+    connection.inOwnerId = boundary.stageId
+
+    connection.outOwner = logic
+    connection.id = -1 // Will be filled later
+    connection.outOwnerId = logic.stageId
+    connection.outHandler = logic.handlers(logic.inCount + out.id).asInstanceOf[OutHandler]
+    logic.portToConn(logic.inCount + out.id) = connection
+
+    boundary.publisher
+  }
+
+  override def takePublisher(slot: Int, publisher: Publisher[Any]): Unit = {
+    // TODO: proper input buffer sizes from attributes
+    val connection = conn(slot)
+    // TODO: proper input port debug string (currently prints the stage)
+    val boundary = new BatchingActorInputBoundary(16, shell, publisher, connection.inOwner.toString)
+    logics.add(boundary)
+    boundary.stageId = logics.size() - 1
+
+    boundary.portToConn(boundary.out.id + boundary.inCount) = connection
+    connection.outHandler = boundary.handlers(0).asInstanceOf[OutHandler]
+    connection.outOwner = boundary
+    connection.outOwnerId = boundary.stageId
+  }
+
+  override def onIslandReady(): Unit = {
+
+    val totalConnections = maxConnections + outConnections.size + 1
```

Comment:
true, I can track this in a var instead. I guess this should be an ArrayList anyway, to avoid linked structures.